src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+			uint8_t palID = mappings[iter->protoPaletteID] & 7;
+			attrmapOutput->sputc(palID | bank << 3); // The other flags are all 0
+			++iter;
+		}
+		++tileID;
+	}
+}
+
+} // namespace unoptimized
+
+static uint8_t flip(uint8_t byte) {
+	// To flip all the bits, we'll flip both nibbles, then each nibble half, etc.
+	byte = (byte & 0x0F) << 4 | (byte & 0xF0) >> 4;
+	byte = (byte & 0x33) << 2 | (byte & 0xCC) >> 2;
+	byte = (byte & 0x55) << 1 | (byte & 0xAA) >> 1;
+	return byte;
+}
+
+class TileData {
+	// TODO: might want to switch to `std::byte` instead?
+	std::array<uint8_t, 16> _data;
+	// The hash is a bit lax: it's the XOR of all lines, and every other nibble is identical
+	// if horizontal mirroring is in effect. It should still be a reasonable tie-breaker in
+	// non-pathological cases.
+	uint16_t _hash;
+public:
+	mutable size_t tileID;
+
+	TileData(Png::TilesVisitor::Tile const &tile, Palette const &palette) : _hash(0) {
+		for (uint32_t y = 0; y < 8; ++y) {
+			uint16_t bitplanes = 0;
+			for (uint32_t x = 0; x < 8; ++x) {
+				bitplanes <<= 1;
+				uint8_t index = palette.indexOf(tile.pixel(x, y));
+				if (index & 1) {
+					bitplanes |= 1;
+				}
+				if (index & 2) {
+					bitplanes |= 0x100;
+				}
+			}
+			_data[y * 2] = bitplanes & 0xFF;
+			_data[y * 2 + 1] = bitplanes >> 8;
+
+			// Update the hash
+			_hash ^= bitplanes;
+			if (options.allowMirroring) {
+				// Count the line itself as mirrorred; vertical mirroring is
+				// already taken care of because the symmetric line will be XOR'd
+				// the same way. (...which is a problem, but probably benign.)
+				_hash ^= flip(bitplanes >> 8) << 8 | flip(bitplanes & 0xFF);
+			}
+		}
+	}
+
+	auto const &data() const { return _data; }
+	uint16_t hash() const { return _hash; }
+
+	enum MatchType {
+		NOPE,
+		EXACT,
+		HFLIP,
+		VFLIP,
+		VHFLIP,
+	};
+	MatchType tryMatching(TileData const &other) const {
+		if (std::equal(_data.begin(), _data.end(), other._data.begin()))
+			return MatchType::EXACT;
+
+		if (options.allowMirroring) {
+			// TODO
+		}
+
+		return MatchType::NOPE;
+	}
+	friend bool operator==(TileData const &lhs, TileData const &rhs) {
+		return lhs.tryMatching(rhs) != MatchType::NOPE;
+	}
+};
+
+template<>
+struct std::hash<TileData> {
+	std::size_t operator()(TileData const &tile) const { return tile.hash(); }
+};
+
+namespace optimized {
+
+struct UniqueTiles {
+	std::unordered_set<TileData> tileset;
+	std::vector<TileData const *> tiles;
+
+	UniqueTiles() = default;
+	// Copies are likely to break pointers, so we really don't want those.
+	// Copy elision should be relied on to be more sure that refs won't be invalidated, too!
+	UniqueTiles(UniqueTiles const &) = delete;
+	UniqueTiles(UniqueTiles &&) = default;
+
+	/**
+	 * Adds a tile to the collection, and returns its ID
+	 */
+	std::tuple<uint16_t, TileData::MatchType> addTile(Png::TilesVisitor::Tile const &tile,
+	                                                  Palette const &palette) {
```

Comment:
I'm finding these highly-indented function arguments annoying to read. From work I'm more used to:
```suggestion
	std::tuple<uint16_t, TileData::MatchType> addTile(
		Png::TilesVisitor::Tile const &tile,
		Palette const &palette
	) {
```
But if you explicitly prefer it this way, that's alright.

Reply:
(Another reason I prefer that way, apart from less indentation, is that there's one argument per line, instead of potentially having two or more short arguments together and then long ones on their own, as `dedupTiles` below does.)